### ISSUE TITLE
`from store import *` overriding `__name__` and `__package__` in PY2

### ISSUE THREAD

**Issue by ZanithOne** [renpy/renpy#4077]
I've tried upgrading my game from RenPy 7.3.5 to 7.5.3 and ran into an issue with pickling classes in Names Stores (modules).

Here's a minimal reproducible example. The project was created in 7.3.5 (doesn't seem to matter, though).

```py
﻿init python in named_store_test:

    from store import *

    foo_inst = None

    class Foo:
        def __init__(self, foobar):
            self.foobar = foobar

    def init_foo():
        named_store_test.foo_inst = Foo("bar")

label start:
    $ named_store_test.init_foo()
    "Test start. Foo initialized."
    "Save here."
    "Test read: [named_store_test.foo_inst.foobar]"
    "Test end"
    return
```

When saving at `Save here`, in **RenPy 7.3.5** the state is saved without any issue. You can load the save, the object will be there with the data intact.

When saving at the same point (or anywhere after Foo is initialized) in **RenPy 7.5.3**, the pickle crashes with:

```
Full traceback:
  File "renpy/common/_layout/screen_load_save.rpym", line 35, in script
    $ ui.interact()
  File "renpy/ast.py", line 1131, in execute
    renpy.python.py_exec_bytecode(self.code.bytecode, self.hide, store=self.store)
  File "renpy/python.py", line 1061, in py_exec_bytecode
    exec(bytecode, globals, locals)
  File "renpy/common/_layout/screen_load_save.rpym", line 35, in <module>
    $ ui.interact()
  File "renpy/ui.py", line 299, in interact
    rv = renpy.game.interface.interact(roll_forward=roll_forward, **kwargs)
  File "renpy/display/core.py", line 3377, in interact
    repeat, rv = self.interact_core(preloads=preloads, trans_pause=trans_pause, pause=pause, pause_start=pause_start, pause_modal=pause_modal, **kwargs) # type: ignore
  File "renpy/display/core.py", line 4258, in interact_core
    rv = root_widget.event(ev, x, y, 0)
  File "renpy/display/layout.py", line 1175, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/transition.py", line 53, in event
    return self.new_widget.event(ev, x, y, st) # E1101
  File "renpy/display/layout.py", line 1175, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/layout.py", line 1175, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/screen.py", line 743, in event
    rv = self.child.event(ev, x, y, st)
  File "renpy/display/layout.py", line 1175, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/layout.py", line 1399, in event
    rv = super(Window, self).event(ev, x, y, st)
  File "renpy/display/layout.py", line 279, in event
    rv = d.event(ev, x - xo, y - yo, st)
  File "renpy/display/layout.py", line 1175, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/layout.py", line 1175, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/behavior.py", line 1073, in event
    return handle_click(self.clicked)
  File "renpy/display/behavior.py", line 1008, in handle_click
    rv = run(action)
  File "renpy/display/behavior.py", line 322, in run
    new_rv = run(i, *args, **kwargs)
  File "renpy/display/behavior.py", line 329, in run
    return action(*args, **kwargs)
  File "renpy/common/00action_file.rpy", line 383, in __call__
    renpy.save(fn, extra_info=save_name)
  File "renpy/loadsave.py", line 403, in save
    reraise(t, e, tb)
  File "renpy/loadsave.py", line 389, in save
    dump((roots, renpy.game.log), logf)
  File "renpy/compat/pickle.py", line 53, in dump
    pickle.dump(o, f,PROTOCOL)
  File "/home/tom/ab/renpy-build/tmp/install.linux-x86_64/lib/python2.7/pickle.py", line 1376, in dump
  File "/home/tom/ab/renpy-build/tmp/install.linux-x86_64/lib/python2.7/pickle.py", line 224, in dump
  File "/home/tom/ab/renpy-build/tmp/install.linux-x86_64/lib/python2.7/pickle.py", line 286, in save
  File "/home/tom/ab/renpy-build/tmp/install.linux-x86_64/lib/python2.7/pickle.py", line 554, in save_tuple
  File "/home/tom/ab/renpy-build/tmp/install.linux-x86_64/lib/python2.7/pickle.py", line 286, in save
  File "/home/tom/ab/renpy-build/tmp/install.linux-x86_64/lib/python2.7/pickle.py", line 655, in save_dict
  File "/home/tom/ab/renpy-build/tmp/install.linux-x86_64/lib/python2.7/pickle.py", line 687, in _batch_setitems
  File "/home/tom/ab/renpy-build/tmp/install.linux-x86_64/lib/python2.7/pickle.py", line 331, in save
  File "/home/tom/ab/renpy-build/tmp/install.linux-x86_64/lib/python2.7/pickle.py", line 396, in save_reduce
  File "/home/tom/ab/renpy-build/tmp/install.linux-x86_64/lib/python2.7/pickle.py", line 286, in save
  File "/home/tom/ab/renpy-build/tmp/install.linux-x86_64/lib/python2.7/pickle.py", line 754, in save_global
PicklingError: Can't pickle <class 'store.Foo'>: it's not found as store.Foo
```

**Comment by ZanithOne** [comment]
If there's any issue with reproducibility, I can provide additional info here or on RenPy Discord, I post there as Zanith#0414 

I did some more digging and found out the following pattern. RenPy versions **7.4.0 to 7.4.11** crash at the very start of the script when trying to assign the variable:

```
Full traceback:
  File "game/script.rpy", line 15, in script
    $ named_store_test.init_foo()
  File "renpy/ast.py", line 928, in execute
    renpy.python.py_exec_bytecode(self.code.bytecode, self.hide, store=self.store)
  File "renpy/python.py", line 2245, in py_exec_bytecode
    exec(bytecode, globals, locals)
  File "game/script.rpy", line 15, in <module>
    $ named_store_test.init_foo()
  File "game/script.rpy", line 12, in init_foo
    named_store_test.foo_inst = Foo("bar")
NameError: global name 'named_store_test' is not defined

Windows-10-10.0.19041
Ren'Py 7.4.11.2266
```

RenPy versions **7.5.0+** all crash on saving with the same tracebak I posted in the initial message:

```
Full traceback:
  File "renpy/common/_layout/screen_load_save.rpym", line 35, in script
    $ ui.interact()
  File "renpy/ast.py", line 1111, in execute
    renpy.python.py_exec_bytecode(self.code.bytecode, self.hide, store=self.store)
  File "renpy/python.py", line 1049, in py_exec_bytecode
    exec(bytecode, globals, locals)
  File "renpy/common/_layout/screen_load_save.rpym", line 35, in <module>
    $ ui.interact()
  File "renpy/ui.py", line 299, in interact
    rv = renpy.game.interface.interact(roll_forward=roll_forward, **kwargs)
  File "renpy/display/core.py", line 3376, in interact
    repeat, rv = self.interact_core(preloads=preloads, trans_pause=trans_pause, pause=pause, pause_start=pause_start, **kwargs) # type: ignore
  File "renpy/display/core.py", line 4251, in interact_core
    rv = root_widget.event(ev, x, y, 0)
  File "renpy/display/layout.py", line 1175, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/transition.py", line 53, in event
    return self.new_widget.event(ev, x, y, st) # E1101
  File "renpy/display/layout.py", line 1175, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/layout.py", line 1175, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/screen.py", line 743, in event
    rv = self.child.event(ev, x, y, st)
  File "renpy/display/layout.py", line 1175, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/layout.py", line 1399, in event
    rv = super(Window, self).event(ev, x, y, st)
  File "renpy/display/layout.py", line 279, in event
    rv = d.event(ev, x - xo, y - yo, st)
  File "renpy/display/layout.py", line 1175, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/layout.py", line 1399, in event
    rv = super(Window, self).event(ev, x, y, st)
  File "renpy/display/layout.py", line 279, in event
    rv = d.event(ev, x - xo, y - yo, st)
  File "renpy/display/layout.py", line 1175, in event
    rv = i.event(ev, x - xo, y - yo, cst)
  File "renpy/display/layout.py", line 279, in event
    rv = d.event(ev, x - xo, y - yo, st)
  File "renpy/display/behavior.py", line 1054, in event
    return handle_click(self.clicked)
  File "renpy/display/behavior.py", line 989, in handle_click
    rv = run(action)
  File "renpy/display/behavior.py", line 329, in run
    return action(*args, **kwargs)
  File "renpy/common/00action_file.rpy", line 383, in __call__
    renpy.save(fn, extra_info=save_name)
  File "renpy/loadsave.py", line 403, in save
    reraise(t, e, tb)
  File "renpy/loadsave.py", line 389, in save
    dump((roots, renpy.game.log), logf)
  File "renpy/compat/pickle.py", line 51, in dump
    cPickle.dump(o, f, PROTOCOL)
PicklingError: Can't pickle <class 'store.Foo'>: attribute lookup store.Foo failed

Windows-10-10.0.19041 AMD64
Ren'Py 7.5.0.22062402
```

**Comment by Gouvernathor** [comment]
No error on renpy8, but it does occur on the renpy7 stable release.
This is the base problem : in renpy8 the class has the correct name, aka prepended with the store name :
![image](https://user-images.githubusercontent.com/44340603/200289540-bb288f5a-33f4-47e2-8d58-4521bc4cf000.png)
But in renpy7 it's not, it's labelled with "store" instead.
![image](https://user-images.githubusercontent.com/44340603/200289506-35e659d8-1a70-46b5-b70a-37a613b74a7b.png)

**Comment by Gouvernathor** [comment]
Ok, I found the mistake.
When you do `named_store_test.Foo` in the init function, it looks for `named_store_test` _within_ where it's located, aka `named_store_test` itself. You should directly do `global foo_inst` then `foo_inst = Foo()`.
To patch that, you're star-importing all names in `store` into `named_store_test`, hoping that `named_store_test` itself will be in it. That's a gamble, and basically it's bad python to have a submodule import a more global module, and especially to have it hope that the submodule namespace is initialized.
TLDR : we won't ensure the 7.4 error won't strike, for python reasons and not renpy reasons.

However, there is still a problem. The following code works :
```rpy
init python in named_store_test:

    # from store import *

    foo_inst = None

    class Foo:
        def __init__(self, foobar):
            self.foobar = foobar

    def init_foo():
        global foo_inst
        foo_inst = Foo("bar")
```
But as soon as you readd the `from store import *` (which should be benigh, in that case), the error strikes back with the Foo class being incorrectly named `store.Foo`. The following code tests just that :
```rpy
init python in named:
    # from store import *

    class Foo: pass

default fo = None

label start:
    $ fo = named.Foo()
    "Test start. Foo initialized."
    '(You can save here)\nTest read: [fo]\nIt should read "named.Foo" or "store.named.Foo", not "store.Foo".'
    return
```
It works when the line is commented out, but not when it's readded.

**Comment by Booplicate** [comment]
I think it's similar to #3693. Maybe you could just ignore all `__dunder__` members.

**Comment by Gouvernathor** [comment]
That's true, but I doubt the solution will be similar. The fact that it only happens in py2 is weird.
Some new info, this is what happens in py2 :
![image](https://user-images.githubusercontent.com/44340603/200311819-bce7e8f4-fdcc-4dfe-81e1-c4ac0247363a.png)
And yet :
![image](https://user-images.githubusercontent.com/44340603/200312414-4f2a8106-7b5d-4444-b58e-205fb7f5fcc0.png)

**Comment by Gouvernathor** [comment]
I located the problem and found a fix. This is one of the most vicious bugs I ever encountered on renpy.
